dkeyserver/Enclave/enclave.cpp

```diff
@@ -494,3 +494,278 @@ int sgx_set_up_tls_server(char *server_port)
+        goto done;
+    }
+
+    dest_sock.sin_family = AF_INET;
+    dest_sock.sin_port = htons(server_port);
+    dest_sock.sin_addr.s_addr = inet_addr2(server_name);
+    bzero(&(dest_sock.sin_zero), sizeof(dest_sock.sin_zero));
+
+    if (connect(
+            sockfd, (sockaddr *)&dest_sock,
+            sizeof(struct sockaddr)) == -1)
+    {
+        log_d(
+            TLS_SERVER "failed to connect to target server %d:%d (errno=%d)\n",
+            server_name,
+            server_port,
+            errno);
+        ocall_close(&res, sockfd);
+        if (res != 0)
+            log_d(TLS_SERVER "OCALL: error closing socket\n");
+        sockfd = -1;
+        goto done;
+    }
+    log_d(TLS_SERVER "connected to target server %s:%d\n", server_name, server_port);
+
+done:
+    return sockfd;
+}
+
+int get_dkey_from_target(SSL *ssl)
+{
+    unsigned char buf[200];
+    int ret = 1;
+    int error = 0;
+    int len = 0;
+    int bytes_written = 0;
+    int bytes_read = 0;
+    // sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    // Write an GET request to the target server
+    log_d(TLS_SERVER "-----> Write to server:\n");
+    len = snprintf((char *)buf, sizeof(buf) - 1, CLIENT_PAYLOAD);
+
+    while ((bytes_written = SSL_write(ssl, buf, (size_t)len)) <= 0)
+    {
+        error = SSL_get_error(ssl, bytes_written);
+        if (error == SSL_ERROR_WANT_WRITE)
+            continue;
+        log_d(TLS_SERVER "Failed! SSL_write returned %d\n", error);
+        ret = bytes_written;
+        goto done;
+    }
+
+    log_d(TLS_SERVER "%d bytes written\n", bytes_written);
+
+    // Read the HTTP response from target server
+    log_d(TLS_SERVER "<---- Read from server:\n");
+    do
+    {
+        len = sizeof(buf) - 1;
+        memset(buf, 0, sizeof(buf));
+        bytes_read = SSL_read(ssl, buf, (size_t)len);
+
+        if (bytes_read <= 0)
+        {
+            int error = SSL_get_error(ssl, bytes_read);
+            if (error == SSL_ERROR_WANT_READ)
+                continue;
+
+            log_d(TLS_SERVER "Failed! SSL_read returned error=%d\n", error);
+            ret = bytes_read;
+            break;
+        }
+
+        log_d(TLS_SERVER " %d bytes read\n", bytes_read);
+
+        if (bytes_read != SGX_DOMAIN_KEY_SIZE)
+        {
+            log_d(
+                TLS_SERVER  "ERROR: expected reading %lu bytes but only "
+                           "received %d bytes\n",
+                SGX_DOMAIN_KEY_SIZE,
+                bytes_read);
+            ret = bytes_read;
+            break;
+        }
+        else
+        {
+            log_d(TLS_SERVER
+                  " received all the expected data from server\n\n");
+            ret = 0;
+            uint8_t g_domain_key[SGX_DOMAIN_KEY_SIZE] ;
+            uint32_t dk_cipher_len = sgx_calc_sealed_data_size(0, SGX_DOMAIN_KEY_SIZE);
+            uint8_t dk_cipher[dk_cipher_len] = {0};
+            memcpy_s(g_domain_key, SGX_DOMAIN_KEY_SIZE, buf, SGX_DOMAIN_KEY_SIZE);
```

Comment:
buf contains secret, memset_s() it before function return.

Reply:
done.

Reply:
memset_s(buf);

Reply:
done.